File: app/src/main/java/com/codylimber/fieldphenology/data/model/Dataset.kt
package com.codylimber.fieldphenology.data.model

import kotlinx.serialization.Serializable

@Serializable
data class Dataset(
    val metadata: DatasetMetadata,
    val species: List<Species>
)

@Serializable
data class DatasetMetadata(
    val placeName: String,
    val placeId: Int,
    val placeIds: List<Int> = emptyList(),
    val group: String,
    val taxonName: String,
    val taxonIds: List<Int> = emptyList(),
    val totalObs: Int,
    val speciesCount: Int,
    val generatedAt: String,
    val minObs: Int = 1,
    val qualityGrade: String = "research",
    val maxPhotos: Int = 3
)

@Serializable
data class Species(
    val taxonId: Int,
    val scientificName: String,
    val commonName: String = "",
    val totalObs: Int,
    val rarity: String,
    val peakWeek: Int,
    val firstWeek: Int,
    val lastWeek: Int,
    val periodCount: Int,
    val description: String = "",
    val conservationStatus: String? = null,
    val conservationStatusName: String? = null,
    val family: String? = null,
    val familyScientific: String? = null,
    val order: String? = null,
    val orderScientific: String? = null,
    val photos: List<SpeciesPhoto> = emptyList(),
    val weekly: List<WeeklyEntry> = emptyList()
)

@Serializable
data class SpeciesPhoto(
    val file: String,
    val attribution: String? = null,
    val license: String? = null
)

@Serializable
data class WeeklyEntry(
    val week: Int,
    val n: Int,
    val relAbundance: Float
)

enum class SpeciesStatus {
    PEAK, ACTIVE, EARLY, LATE, INACTIVE;

    companion object {
        fun classify(species: Species, currentWeek: Int): SpeciesStatus {
            val entry = species.weekly.find { it.week == currentWeek }
            if (entry == null || entry.n == 0) return INACTIVE
            return when {
                entry.relAbundance >= 0.8f -> PEAK
                entry.relAbundance >= 0.2f -> ACTIVE
                currentWeek < species.peakWeek -> EARLY
                else -> LATE
            }
        }
    }
}

enum class SortMode {
    LIKELIHOOD, PEAK_DATE, NAME, TAXONOMY
}

enum class ObservationFilter {
    ALL, OBSERVED, NOT_OBSERVED, FAVORITES
}

File: ios/Manakin/Data/Models/Dataset.swift
import Foundation

struct Dataset: Codable {
    let metadata: DatasetMetadata
    let species: [Species]
}

struct DatasetMetadata: Codable {
    let placeName: String
    let placeId: Int
    let placeIds: [Int]
    let group: String
    let taxonName: String
    let taxonIds: [Int]
    let totalObs: Int
    let speciesCount: Int
    let generatedAt: String
    let minObs: Int
    let qualityGrade: String
    let maxPhotos: Int

    init(
        placeName: String,
        placeId: Int,
        placeIds: [Int] = [],
        group: String,
        taxonName: String,
        taxonIds: [Int] = [],
        totalObs: Int,
        speciesCount: Int,
        generatedAt: String,
        minObs: Int = 10,
        qualityGrade: String = "research",
        maxPhotos: Int = 3
    ) {
        self.placeName = placeName
        self.placeId = placeId
        self.placeIds = placeIds
        self.group = group
        self.taxonName = taxonName
        self.taxonIds = taxonIds
        self.totalObs = totalObs
        self.speciesCount = speciesCount
        self.generatedAt = generatedAt
        self.minObs = minObs
        self.qualityGrade = qualityGrade
        self.maxPhotos = maxPhotos
    }
}

struct Species: Codable {
    let taxonId: Int
    let scientificName: String
    let commonName: String
    let totalObs: Int
    let rarity: String
    let peakWeek: Int
    let firstWeek: Int
    let lastWeek: Int
    let periodCount: Int
    let description: String
    let conservationStatus: String?
    let conservationStatusName: String?
    let family: String?
    let familyScientific: String?
    let order: String?
    let orderScientific: String?
    let photos: [SpeciesPhoto]
    let weekly: [WeeklyEntry]

    init(
        taxonId: Int,
        scientificName: String,
        commonName: String = "",
        totalObs: Int,
        rarity: String,
        peakWeek: Int,
        firstWeek: Int,
        lastWeek: Int,
        periodCount: Int,
        description: String = "",
        conservationStatus: String? = nil,
        conservationStatusName: String? = nil,
        family: String? = nil,
        familyScientific: String? = nil,
        order: String? = nil,
        orderScientific: String? = nil,
        photos: [SpeciesPhoto] = [],
        weekly: [WeeklyEntry] = []
    ) {
        self.taxonId = taxonId
        self.scientificName = scientificName
        self.commonName = commonName
        self.totalObs = totalObs
        self.rarity = rarity
        self.peakWeek = peakWeek
        self.firstWeek = firstWeek
        self.lastWeek = lastWeek
        self.periodCount = periodCount
        self.description = description
        self.conservationStatus = conservationStatus
        self.conservationStatusName = conservationStatusName
        self.family = family
        self.familyScientific = familyScientific
        self.order = order
        self.orderScientific = orderScientific
        self.photos = photos
        self.weekly = weekly
    }
}

struct SpeciesPhoto: Codable {
    let file: String
    let attribution: String?
    let license: String?

    init(file: String, attribution: String? = nil, license: String? = nil) {
        self.file = file
        self.attribution = attribution
        self.license = license
    }
}

struct WeeklyEntry: Codable {
    let week: Int
    let n: Int
    let relAbundance: Float
}

enum SpeciesStatus {
    case peak, active, early, late, inactive

    static func classify(species: Species, currentWeek: Int) -> SpeciesStatus {
        guard let entry = species.weekly.first(where: { $0.week == currentWeek }),
              entry.n > 0 else {
            return .inactive
        }
        if entry.relAbundance >= 0.8 {
            return .peak
        } else if entry.relAbundance >= 0.2 {
            return .active
        } else {
            // Handle year-wrapping: compute shortest distance to peak in either direction
            let forwardToPeak = (species.peakWeek - currentWeek + 53) % 53
            let backFromPeak = (currentWeek - species.peakWeek + 53) % 53
            return forwardToPeak <= backFromPeak ? .early : .late
        }
    }
}

enum SortMode {
    case likelihood, peakDate, name, taxonomy
}

enum ObservationFilter {
    case all, observed, notObserved, favorites
}
